internal/freelist/freelist_test.go

```diff
@@ -266,6 +356,250 @@ func TestFreelist_write(t *testing.T) {
+	freelist := newTestFreelist()
+	freelist.Init([]common.Pgid{})
+	freelist.Free(common.Txid(10), common.NewPage(10, common.LeafPageFlag, 0, 2))
+	freelist.Free(common.Txid(11), common.NewPage(20, common.LeafPageFlag, 0, 2))
+	freelist.Free(common.Txid(12), common.NewPage(30, common.LeafPageFlag, 0, 2))
+	requirePages(t, freelist, common.Pgids{}, common.Pgids{10, 11, 12, 20, 21, 22, 30, 31, 32})
+
+	freelist.AddReadonlyTXID(11)
+	freelist.ReleasePendingPages()
+	requirePages(t, freelist, common.Pgids{10, 11, 12}, common.Pgids{20, 21, 22, 30, 31, 32})
+
+	// this should be a no-op, as we still have a read TX open with id 11
+	freelist.AddReadonlyTXID(12)
+	freelist.ReleasePendingPages()
+	requirePages(t, freelist, common.Pgids{10, 11, 12}, common.Pgids{20, 21, 22, 30, 31, 32})
+
+	// now 12 should be the latest
+	freelist.RemoveReadonlyTXID(11)
+	freelist.AddReadonlyTXID(12)
+	freelist.ReleasePendingPages()
+	requirePages(t, freelist, common.Pgids{10, 11, 12, 20, 21, 22}, common.Pgids{30, 31, 32})
+
+	// 12 was registered twice, so we also have to remove it twice to have an effect
+	freelist.RemoveReadonlyTXID(12)
+	freelist.AddReadonlyTXID(13)
+	freelist.ReleasePendingPages()
+	requirePages(t, freelist, common.Pgids{10, 11, 12, 20, 21, 22}, common.Pgids{30, 31, 32})
+
+	freelist.RemoveReadonlyTXID(12)
+	freelist.ReleasePendingPages()
+	requirePages(t, freelist, common.Pgids{10, 11, 12, 20, 21, 22, 30, 31, 32}, common.Pgids{})
+}
+
+// tests the reloading from another physical page
+func TestFreelist_E2E_Reload(t *testing.T) {
+	freelist := newTestFreelist()
+	freelist.Init([]common.Pgid{})
+	freelist.Free(common.Txid(2), common.NewPage(5, common.LeafPageFlag, 0, 1))
+	freelist.Free(common.Txid(2), common.NewPage(8, common.LeafPageFlag, 0, 0))
+	freelist.ReleasePendingPages()
+	requirePages(t, freelist, common.Pgids{5, 6, 8}, common.Pgids{})
+	buf := make([]byte, 4096)
+	p := common.LoadPage(buf)
+	freelist.Write(p)
+
+	freelist.Free(common.Txid(3), common.NewPage(3, common.LeafPageFlag, 0, 1))
+	freelist.Free(common.Txid(3), common.NewPage(10, common.LeafPageFlag, 0, 2))
+	requirePages(t, freelist, common.Pgids{5, 6, 8}, common.Pgids{3, 4, 10, 11, 12})
+
+	otherBuf := make([]byte, 4096)
+	px := common.LoadPage(otherBuf)
+	freelist.Write(px)
+
+	loadFreeList := newTestFreelist()
+	loadFreeList.Init([]common.Pgid{})
+	loadFreeList.Read(px)
+	requirePages(t, loadFreeList, common.Pgids{3, 4, 5, 6, 8, 10, 11, 12}, common.Pgids{})
+	// restore the original freelist again
+	loadFreeList.Reload(p)
+	requirePages(t, loadFreeList, common.Pgids{5, 6, 8}, common.Pgids{})
+
+	// reload another page with different free pages to test we are deduplicating the free pages with the pending ones correctly
+	freelist = newTestFreelist()
+	freelist.Init([]common.Pgid{})
+	freelist.Free(common.Txid(5), common.NewPage(5, common.LeafPageFlag, 0, 4))
+	freelist.Reload(p)
+	requirePages(t, freelist, common.Pgids{}, common.Pgids{5, 6, 7, 8, 9})
```

Comment:
We can keep it as it's for now. But actually reload is only used in rollback case, so there should be a rollback operation. We can revisit this when we refactor the interface.

Reply:
ack